pkg/ssh/suftp.go

```diff
@@ -0,0 +1,169 @@
+	// sudo -S dictates that password should be read from stdin
+	// sudo -u specifies the user
+	// sh -c 'cmd' passes cmd to sh
+	// >&2 echo "%s" echos promptsucess to stderr before starting sftp server ( see https://stackoverflow.com/a/23550347 )
+	// & %s is the final command to be executed is sudo is successful, which is the path to the sftp-server binary. It is a list of possible binary paths and it will pick the first one that exists. At the end, it will look in $PATH.
+	// || sh -c '>&2 echo "%s" & exit 1' -> runs if sudo failed and ensures that prompterror is written to stderr before erroring
+	// In total this means that we can read from stderr and write to stdin to
+	cmd := fmt.Sprintf(
+		`sudo -p '%s' -S sh -c '>&2 echo "%s" & %s' || sh -c '>&2 echo "%s" & exit 1'`,
+		promptpwd+"\n",
+		promptsuccess,
+		sftpServerPath,
+		prompterror,
+	)
+
+	stdin, err := s.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	stdout, err := s.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	stderr, err := s.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.Start(cmd)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Sudo might output a lecture, so looping for either password, error or success prompt
+	prompt, err := getPrompt(stderr)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get err prompt")
+	}
+
+	switch prompt {
+	case prompterror:
+		return nil, errors.New("sudo failed")
+	case promptpwd:
+		// ignore any writer error
+		_, _ = stdin.Write([]byte(sudopwd + "\n"))
+	case promptsuccess:
+		return sftp.NewClientPipe(stdout, stdin, opts...)
+	}
+
+	// Second read is only after pwd has been entered.
+	// If it returns pwd prompt again, it means that it was the wrong password.
+	// Error means the user is not allowed to sudo or something else went wrong.
+	prompt, err = getPrompt(stderr)
+	if err != nil {
+		return nil, errors.Wrap(err, "second stderr read failed")
+	}
+
+	switch prompt {
+	case prompterror:
+		return nil, errors.New("sudo failed or no sudo rights")
+	case promptpwd:
+		return nil, fmt.Errorf("wrong sudo password")
+	}
+
+	return sftp.NewClientPipe(stdout, stdin, opts...)
+}
+
+// getPrompt is a handy method for reading a prompt from stderr
+// prompt should be at the end of the read, minus a newline
+func getPrompt(rd io.Reader) (string, error) {
+	buf := make([]byte, 2048)
+
+	var prompt string
+
+	for {
+		n, err := rd.Read(buf)
+		if err != nil {
+			return "", err
+		}
+
+		if n <= promptlength+1 {
+			prompt = string(buf[:n-1])
+		} else {
+			prompt = string(buf[n-promptlength-1 : n-1])
+		}
+
+		if prompt == prompterror || prompt == promptpwd || prompt == promptsuccess {
+			return prompt, nil
+		}
+	}
+}
+
+func getSftpServerPath(conn *ssh.Client, sudopwd string) (string, error) {
+	ses, err := conn.NewSession()
+	if err != nil {
+		return "", err
+	}
+	defer ses.Close()
+
+	cmd := fmt.Sprintf(`sudo -S grep -oP "Subsystem\s+sftp\s+\K.*" %s`, sshdConfigPath)
```

Comment:
sudo is not required for this command

Reply:
> sudo is not required for this command

```
[root@node1 ~]# ll /etc/ssh/sshd_config 
-rw-------. 1 root root 3905 Aug  3  2022 /etc/ssh/sshd_config
```

Reply:
weird. the permission of this file on most of our servers are 644.